### PR TITLE
Item 18 reference inputs

### PR DESCRIPTION
**Pull request by rustamlabs**

### CODE REVIEW

File: src/interactions/serialization/txOther.ts [hunk 1]
```diff
@@ -109,3 +109,12 @@ export function serializeTotalCollateral(
         uint64_to_buf(totalCollateral),
     ])
 }
+
+export function serializeReferenceInputs(
```
Comment (janmazak) [inline src/interactions/serialization/txOther.ts:113]:
```suggestion
export function serializeReferenceInput(
```

Comment (janmazak) [reply src/interactions/serialization/txOther.ts:113]: I don't see a function to serialize collateral inputs. But it seems preferable to me to only have a single function for serializing ParsedInput used in 3 places --- that's how the APDU parsing side is written, anyway.

File: src/interactions/signTx.ts [hunk 2]
```diff
@@ -623,6 +624,20 @@ function* signTx_addTotalCollateral(
     })
 }
 
+function* signTx_addTotalReferenceInputs(
+    referenceInputs: ParsedInput
```
Comment (janmazak) [inline src/interactions/signTx.ts:628]: This does not seem right --- why is the type only a single input?

Comment (rustamlabs) [reply src/interactions/signTx.ts:628]: this method called for every input one by on, similar to how  add collateral  is being added

File: src/interactions/signTx.ts [hunk 3]
```diff
@@ -623,6 +624,20 @@ function* signTx_addTotalCollateral(
     })
 }
 
+function* signTx_addReferenceInput(
+    referenceInputs: ParsedInput
```
Comment (janmazak) [inline src/interactions/signTx.ts:628]:
```suggestion
    referenceInput: ParsedInput
```

File: src/interactions/signTx.ts [hunk 4]
```diff
@@ -971,6 +986,13 @@ export function* signTransaction(version: Version, request: ParsedSigningRequest
         yield* signTx_addTotalCollateral(tx.totalCollateral)
     }
 
+    // totalCollateral
```
Comment (janmazak) [inline src/interactions/signTx.ts:989]: misplaced comment

File: test/integration/__fixtures__/signTxPlutus.ts [hunk 5]
```diff
@@ -466,4 +466,52 @@ export const testsBabbage: SignTxTestcase[] = [
             auxiliaryDataSupplement: null,
         },
     },
+    {
+        // tx does not contain any Plutus elements, but should be accepted (differs only in UI)
+        testname: "Sign tx with change output (Plutus) and Reference Input MAP",
```
Comment (janmazak) [inline test/integration/__fixtures__/signTxPlutus.ts:471]: What does "MAP" mean here? We use arrays for inputs.

**Review by janmazak — COMMENTED** [review]
Take also a look at reject tests. They more-or-less copy security policies --- e.g. if you added some restrictions on tx signing mode in security policies for reference inputs, we need
1. add the proper validation to ledgerjs
2. add reject tests to ledgerjs checking the validation works the same in both ledgerapp and ledgerjs

**Review by janmazak — APPROVED** [review]

**Pull request merged**